TST, BUG: Use python3.6-dbg.

charris opened the pull request:
Pytest-5.1.0 is segfaults on python3.5-dbg. Because we will be dropping 3.5 for NumPy 1.18 we might as well use a later Python version for the debug testing instead of pinning the pytest version. That way there will be no need to unpin pytest when it gets fixed.

EDIT:

Both pytest-5.0.1 and pytest-5.1.0 abort when passed the `--durations` flag on python3.6-dbg, so temporarily disable that flag pending a fix and leave the pytest version unpinned until then.    

<!-- Please be sure you are following the instructions in the dev guidelines
http://www.numpy.org/devdocs/dev/development_workflow.html
-->

<!-- We'd appreciate it if your commit message is properly formatted
http://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message
-->

eric-wieser commented:
Since 3.5 is now our oldest version, should we move our debug build to use a newer version?

charris commented:
I want to keep 3.5 for a while because 1.17 supports it. When 3.8 comes out we can drop it. That should be 2019-10-21 if all goes well, which is early enough that we should be able to release 1.18 with 3.8 wheels.

charris committed:
TST: Pin pytest and use python3.6-dbg.

Pytest-5.1.0 segfaults on python3.5-dbg. Because we will be
dropping 3.5 for NumPy 1.18 we might as well use Python3.6
for the debug testing. That cures the previous segfault,
but exposes another so we also pin pytest to 5.0.1.

charris commented:
Checked python-3.6-dbg and it works. It even passes all its own tests, which wasn't the case with 3.5. Since I have a local debug version of 3.5 I can use to chase down the pytest bug, simply going to 3.6 looks like the easiest long term solution. I'll push a commit for that.

charris commented:
Curious, it still tests with 3.5 even though 3.6 is specified. Hmm... 3.5 is the xenial default python3.

charris commented:
So we need bionic for 3.6 . And... there is a different segfault with python3.6-dbg that really looks like pytest. Results [here](https://travis-ci.org/numpy/numpy/builds/573298172?utm_source=github_status&utm_medium=notification)

mattip commented:
It seems pytest breaks our builds often enough that pinning is the best route forward. Maybe we can set up one build to try latest pytest + oldest python debug and set it so test failure does not fail the checks, but we can do that offline as well once every couple of months or so

mattip commented:
The hypothesis project pins versions and upgrades the versions via a weekly (?) [PR](https://github.com/HypothesisWorks/hypothesis/pull/2074) to make sure nothing breaks

charris commented:
@mattip Yep, I think that is the way to go, it was the original fix in this series :0 I'd like to get 3.6 working though, we will need it when we drop 3.5. I think pinning will work with 3.6 now that I've removed the package that is no longer needed and doesn't exist in bionic. I'm going to play a bit with the new segfault so I help the pytest folks fix things.

charris commented:
Curiously, pytest-5.0.1 *also* segfaults on 3.6-dbg. It is the `--duration` flag that brings it about.

pv commented:
Somebody maybe should lobby pytest to start testing on python debug builds...

eric-wieser commented:
Could https://bugs.python.org/issue35560 be relevant?

charris committed:
TST: Disable `--durations` flag for pytest.

This is (hopefully) temporary. Current versions of pytest abort
when this flag is passed and the tests run with python3.6-dbg.

mattip commented on tools/travis-before-install.sh:
```diff
@@ -36,6 +36,6 @@ fi
 
 pip install --upgrade pip setuptools
-pip install pytz cython pytest==5.0.1
+pip install pytz cython pytest
```
I still think pinning is preferable, and we can periodically issue a PR to update the version in a controlled way

charris replied:
The problem is, we don't know of any working versions of pytest :) I've opened a task issue to fix that when there is a working version.

For the NumPy 1.16.x and 1.17.x branches I will pin the pytest version to 5.0.1 since we know that works with python3.5-dbg. Or at least the problem goes unreported

eric-wieser replied:
My fear is that unless we pin everything in one place, we will forget we ever pinned it.

charris replied:
I have a task for that with 1.18.0 milestone, see #14293.

charris commented:
> Somebody maybe should lobby pytest to start testing on python debug builds...

They are going to start doing that, there are now three new pytest issues involving debug python versions.

charris commented:
> Could https://bugs.python.org/issue35560 be relevant?

Yep, that looks like the problem. Unfortunately they didn't backport the fix to python3.6. I wonder if pytest can fix their format statement?

charris commented:
I'm going to put this in for now.  If no fixes are forthcoming, we can make disabling `--durations` debug specific.
